### PR TITLE
[Android] Support setting badge count

### PR DESCRIPTION
**Pull request by igrayson**

### CODE REVIEW

File: component/index.android.js [hunk 1]
```diff
@@ -41,6 +41,13 @@ NotificationsComponent.prototype.scheduleLocalNotification = function(details: O
 	RNPushNotification.scheduleLocalNotification(details);
 };
 
+NotificationsComponent.prototype.setApplicationIconBadgeNumber = function(number: number) {
+       if (!RNPushNotification.setApplicationIconBadgeNumber) {
+               return;
+       }
```
Comment (igrayson) [inline component/index.android.js]: This check is for backwards compatibility. We use CodePush to push javascript changes ahead of native (Play store) updates, and this lets us do that.

**Comment by kfiroo** [comment]
@igrayson This looks like a much better implementation then my PR!
Should I close mine and we can continue the discussion here?

**Comment by igrayson** [comment]
Sounds good.

**Comment by npomfret** [comment]
I'll look at merging this - are we intending to do with this one instead of https://github.com/zo0r/react-native-push-notification/pull/168 ?

**Comment by kfiroo** [comment]
@npomfret I think yes, this is a better implementation

**Comment by npomfret** [comment]
I tried it out on a THL device and got the following error:

```
E/ShortcutBadger: Unable to execute badge
                                                              me.leolin.shortcutbadger.ShortcutBadgeException: Unable to execute badge
                                                                  at me.leolin.shortcutbadger.ShortcutBadger.applyCountOrThrow(ShortcutBadger.java:77)
                                                                  at me.leolin.shortcutbadger.ShortcutBadger.applyCount(ShortcutBadger.java:52)
                                                                  at com.dieam.reactnativepushnotification.helpers.ApplicationBadgeHelper.tryAutomaticBadge(ApplicationBadgeHelper.java:50)
                                                                  at com.dieam.reactnativepushnotification.helpers.ApplicationBadgeHelper.setApplicationIconBadgeNumber(ApplicationBadgeHelper.java:44)
                                                                  at com.dieam.reactnativepushnotification.modules.RNPushNotification.setApplicationIconBadgeNumber(RNPushNotification.java:239)
                                                                  at java.lang.reflect.Method.invoke(Native Method)
                                                                  at java.lang.reflect.Method.invoke(Method.java:372)
                                                                  at com.facebook.react.bridge.BaseJavaModule$JavaMethod.invoke(BaseJavaModule.java:319)
                                                                  at com.facebook.react.cxxbridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:139)
                                                                  at com.facebook.react.bridge.queue.NativeRunnable.run(Native Method)
                                                                  at android.os.Handler.handleCallback(Handler.java:815)
                                                                  at android.os.Handler.dispatchMessage(Handler.java:104)
                                                                  at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:31)
                                                                  at android.os.Looper.loop(Looper.java:194)
                                                                  at com.facebook.react.bridge.queue.MessageQueueThreadImpl$3.run(MessageQueueThreadImpl.java:196)
                                                                  at java.lang.Thread.run(Thread.java:818)
                                                               Caused by: me.leolin.shortcutbadger.ShortcutBadgeException: unable to resolve intent: Intent { act=android.intent.action.BADGE_COUNT_UPDATE (has extras) }
                                                                  at me.leolin.shortcutbadger.impl.DefaultBadger.executeBadge(DefaultBadger.java:33)
                                                                  at me.leolin.shortcutbadger.ShortcutBadger.applyCountOrThrow(ShortcutBadger.java:75)
                                                                  at me.leolin.shortcutbadger.ShortcutBadger.applyCount(ShortcutBadger.java:52) 
                                                                  at com.dieam.reactnativepushnotification.helpers.ApplicationBadgeHelper.tryAutomaticBadge(ApplicationBadgeHelper.java:50) 
                                                                  at com.dieam.reactnativepushnotification.helpers.ApplicationBadgeHelper.setApplicationIconBadgeNumber(ApplicationBadgeHelper.java:44) 
                                                                  at com.dieam.reactnativepushnotification.modules.RNPushNotification.setApplicationIconBadgeNumber(RNPushNotification.java:239) 
                                                                  at java.lang.reflect.Method.invoke(Native Method) 
                                                                  at java.lang.reflect.Method.invoke(Method.java:372) 
                                                                  at com.facebook.react.bridge.BaseJavaModule$JavaMethod.invoke(BaseJavaModule.java:319) 
                                                                  at com.facebook.react.cxxbridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:139) 
                                                                  at com.facebook.react.bridge.queue.NativeRunnable.run(Native Method) 
                                                                  at android.os.Handler.handleCallback(Handler.java:815) 
                                                                  at android.os.Handler.dispatchMessage(Handler.java:104) 
                                                                  at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:31) 
                                                                  at android.os.Looper.loop(Looper.java:194) 
                                                                  at com.facebook.react.bridge.queue.MessageQueueThreadImpl$3.run(MessageQueueThreadImpl.java:196) 
                                                                  at java.lang.Thread.run(Thread.java:818) 
```

Looks like someone else has the same [issue](https://github.com/leolin310148/ShortcutBadger/issues/109), strangely the suggestion is to comment out some of the code in the library.

**Comment by igrayson** [comment]
As I understand ShortcutBadger's structure, this is intended behavior, and
motivated us to have `ApplicationBadgeHelper` remember whether a badging
failed, and never try again.

The impact as I understand it (please confirm) is one logcat error on your
app's first attempt to set a badge, on a phone with a launcher that does
not support badges.

**Comment by igrayson** [comment]
This could probably be improved, in `ApplicationBadgeHelper`, by using and
catching `applyCountOrThrow` instead of `applyCount`.

**Comment by npomfret** [comment]
The behaviour is as you've described.  But I have no badge - and so can't really test it.  I'll see if I can try a different device.

**Comment by npomfret** [comment]
..and is it necessary to add stuff to the manifest?  I saw someone on stack overflow say they were adding these lines:

```
    <!--for android badge-->
    <uses-permission android:name="com.android.launcher.permission.READ_SETTINGS"/>
    <uses-permission android:name="com.android.launcher.permission.WRITE_SETTINGS"/>

    <!--for Samsung badge-->
    <uses-permission android:name="com.sec.android.provider.badge.permission.READ"/>
    <uses-permission android:name="com.sec.android.provider.badge.permission.WRITE"/>

    <!--for htc badge-->
    <uses-permission android:name="com.htc.launcher.permission.READ_SETTINGS"/>
    <uses-permission android:name="com.htc.launcher.permission.UPDATE_SHORTCUT"/>

    <!--for sony badge-->
    <uses-permission android:name="com.sonyericsson.home.permission.BROADCAST_BADGE"/>

    <!--for apex badge-->
    <uses-permission android:name="com.anddoes.launcher.permission.UPDATE_COUNT"/>

```

**Comment by igrayson** [comment]
It shouldn't be necessary. ShortcutBadger brings its own
`AndroidManifest.xml`

**Comment by npomfret** [comment]
I don't have a physical device that can test this - is anyone able to verify it works?

**Comment by npomfret** [comment]
Finally got this sunning on a Samsung - works like a charm

**Comment by bwoodlt** [comment]
@npomfret How were you able to fix this? Having the same issue

**Pull request merged**